src/pyodide/internal/builtin_wrappers.js

```diff
@@ -0,0 +1,63 @@
+import { default as Eval } from "pyodide-internal:eval";
+
+const EXPECTED_CALLEE =
+  "    at convertJsFunctionToWasm (pyodide-internal:pyodide-bundle/pyodide.asm";
+
+function checkCallee() {
+  const stack = new Error().stack;
+  // First line is Error:, 2nd line is this call frame, 3rd line is
+  // `new WebAssembly.Module()`, 4th line is callee.
+  let calleeLine = stack.split("\n")[3];
+  // Remove position info from line
+  const secondToLastColonPos = calleeLine.lastIndexOf(":", calleeLine.lastIndexOf(":") - 1);
+  calleeLine = calleeLine.slice(0, secondToLastColonPos);
+  if (calleeLine !== EXPECTED_CALLEE) {
+    console.warn("$" + calleeLine + "$");
+    throw new Error();
+  }
+}
```

Comment:
@dom96 see here for eval safety check

Reply:
I guess to make sure that this actually works, we'd have to do a security review to check that it's not possible for a user to make a module whose name renders as `pyodide-internal:pyodide-bundle/pyodide.asm`...

Reply:
Well the point of this is that I'm concerned that global scope symbols could somehow leak from `pyodide.asm.js`. This `WebAssembly` patch is itself a global symbol. I don't *think* there's anyway for user code to get to it, but it's best to be cautious. OTOH, if we permit calling the function on anyone who has a symbol in global scope it kind of feels like the same check again. But perhaps from C++ we have some stronger way to validate the identity of `pyodide.asm.js`.

But I think this check as-is is less fragile than you might think. Someone would have to actually make a file called `pyodide-internal:pyodide-bundle/pyodide.asm.js`, but I don't think users are allowed to make files with names like this. Since I split the colons from the right, even if someone managed to make a function with a name like `createFunction (pyodide-internal:...)`, we would still notice the correct parentheses.

Reply:
OTOH, I wonder if they could spoof it by overriding `Error.prepareStackTrace`. We should consider setting and restoring `Error.prepareStackTrace` here and then we wouldn't have to parse the traceback.

Reply:
Okay, I made an improved version with `prepareStackTrace`.

I think the best thing to do would be to have an actual list of function objects that are allowed to call the function, and then add a C++ function that returns the function up the callstack (forbidden in ordinary strict mode JS) so we can call that to look up the callee and check if it's in a whitelist. That check would be impossible to spoof.